File: project/jimmer-apt/src/main/java/org/babyfish/jimmer/apt/util/GenericParser.java
package org.babyfish.jimmer.apt.util;

import com.squareup.javapoet.*;
import org.babyfish.jimmer.apt.MetaException;

import javax.lang.model.element.TypeElement;
import javax.lang.model.element.TypeParameterElement;
import javax.lang.model.type.*;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class GenericParser {

    private final String name;

    private final TypeElement element;

    private final String superName;

    private final Map<TypeVariable, TypeMirror> replaceMap = new HashMap<>();

    public GenericParser(String name, TypeElement element, String superName) {
        this.name = name.toLowerCase();
        this.element = element;
        this.superName = superName;
    }

    public List<TypeName> parse() {
        try {
            parse((DeclaredType) element.asType());
        } catch (Finished ex) {
            return ex.arguments;
        }
        throw new MetaException(
                element,
                "it does not specify the arguments for \"" +
                        superName +
                        "\""
        );
    }

    private void parse(DeclaredType type) throws Finished {
        TypeElement typeElement = (TypeElement) type.asElement();
        List<? extends TypeMirror> arguments = type.getTypeArguments();
        if (typeElement.getQualifiedName().toString().equals(superName)) {
            if (arguments.isEmpty()) {
                throw new MetaException(
                        this.element,
                        "The " +
                                name +
                                " type \"" +
                                typeElement +
                                "\" does not specify type arguments for \"" +
                                superName +
                                "\""
                );
            }
            throw new Finished(
                    arguments.stream().map(this::resolve).collect(Collectors.toList())
            );
        } else {
            if (!arguments.isEmpty()) {
                List<? extends TypeParameterElement> parameters = typeElement.getTypeParameters();
                int size = arguments.size();
                for (int i = 0; i < size; i++) {
                    replaceMap.put((TypeVariable) parameters.get(i).asType(), arguments.get(i));
                }
            }
            TypeMirror superType = typeElement.getSuperclass();
            if (superType instanceof DeclaredType) {
                parse((DeclaredType) superType);
            }
            for (TypeMirror superItf : typeElement.getInterfaces()) {
                parse((DeclaredType) superItf);
            }
        }
    }

    private TypeName resolve(TypeMirror type) {
        if (type instanceof DeclaredType) {
            DeclaredType declaredType = (DeclaredType) type;
            TypeElement typeElement = (TypeElement) declaredType.asElement();
            ClassName className = ClassName.bestGuess(typeElement.getQualifiedName().toString());
            List<? extends TypeMirror> arguments = declaredType.getTypeArguments();
            if (!arguments.isEmpty()) {
                TypeName[] argTypeNames = new TypeName[arguments.size()];
                for (int i = arguments.size() - 1; i >= 0; --i) {
                    argTypeNames[i] = resolve(arguments.get(i));
                }
                return ParameterizedTypeName.get(className, argTypeNames);
            }
            return className;
        }
        if (type instanceof TypeVariable) {
            return resolve((TypeVariable) type);
        }
        if (type instanceof WildcardType) {
            WildcardType wildcardType = (WildcardType) type;
            if (wildcardType.getSuperBound() != null) {
                return WildcardTypeName.supertypeOf(
                        resolve(wildcardType.getSuperBound())
                );
            }
            return WildcardTypeName.subtypeOf(
                    wildcardType.getExtendsBound() != null ?
                            resolve(wildcardType.getExtendsBound()) :
                            TypeName.OBJECT
            );
        }
        if (type instanceof ArrayType) {
            return ArrayTypeName.of(
                    resolve(((ArrayType)type).getComponentType())
            );
        }
        throw new MetaException(
                element,
                "cannot resolve \"" + type + "\""
        );
    }

    private TypeName resolve(TypeVariable typeVariable) {
        TypeMirror type = replaceMap.get(typeVariable);
        if (type == null) {
            throw new MetaException(
                    element,
                    "cannot resolve \"" + typeVariable + "\""
            );
        }
        return resolve(type);
    }

    private static class Finished extends Exception {

        final List<TypeName> arguments;

        private Finished(List<TypeName> arguments) {
            this.arguments = arguments;
        }
    }
}

File: project/jimmer-ksp/src/main/kotlin/org/babyfish/jimmer/ksp/util/GenericParser.kt
package org.babyfish.jimmer.ksp.util

import com.google.devtools.ksp.symbol.*
import com.squareup.kotlinpoet.ANY
import com.squareup.kotlinpoet.ClassName
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.TypeName
import com.squareup.kotlinpoet.WildcardTypeName
import org.babyfish.jimmer.ksp.MetaException
import java.lang.Exception
import java.lang.IllegalStateException

class GenericParser(
    private val name: String,
    private val declaration: KSClassDeclaration,
    private val superName: String
) {

    private val replaceMap: MutableMap<KSTypeParameter, KSTypeArgument> = mutableMapOf()

    init {
        if (declaration.typeParameters.isNotEmpty()) {
            throw IllegalStateException("\"${declaration.qualifiedName!!.asString()}\" cannot have type parameters");
        }
    }

    fun parse(): Result {
        try {
            parse(declaration.asType(emptyList()))
        } catch (ex: Finished) {
            return ex.result
        }
        throw MetaException(
            declaration,
            "it does not specify the arguments for \"" +
                superName +
                "\""
        )
    }

    private fun parse(type: KSType) {
        if (type.declaration.qualifiedName!!.asString() == superName) {
            if (type.arguments.isEmpty()) {
                throw MetaException(
                    declaration,
                    "it does not specify type argument for \"$superName\""
                )
            }
            throw Finished(Result(type.arguments, type.arguments.map { resolve(it) }))
        } else {
            if (type.arguments.isNotEmpty()) {
                val parameters = type.declaration.typeParameters
                val size = parameters.size
                for (i in 0 until size) {
                    replaceMap[parameters[i]] = type.arguments[i]
                }
            }
            for (superTypeRef in (type.declaration as KSClassDeclaration).superTypes) {
                val superType = superTypeRef.fastResolve()
                parse(superType)
            }
        }
    }

    private fun resolve(type: KSType): TypeName =
        when (type) {
            is KSTypeParameter -> {
                val replaced = replaceMap[type]
                    ?: throw MetaException(
                        declaration,
                        "The type parameter \"$type\" cannot be resolved"
                    )
                resolve(replaced)
            }
            is KSTypeArgument ->
                resolve(type as KSTypeArgument)
            else ->
                ClassName.bestGuess(
                    type.declaration.qualifiedName!!.asString().let {
                        when (it) {
                            "kotlin.collections.MutableIterable" -> "kotlin.collections.Iterable"
                            "kotlin.collections.MutableCollection" -> "kotlin.collections.Collection"
                            "kotlin.collections.MutableList" -> "kotlin.collections.List"
                            "kotlin.collections.MutableSet" -> "kotlin.collections.Set"
                            "kotlin.collections.MutableMap" -> "kotlin.collections.Map"
                            else -> it
                        }
                    }
                ).let {
                    if (type.arguments.isEmpty()) {
                        it
                    } else {
                        it.parameterizedBy(
                            type.arguments.map { a -> resolve(a) }
                        )
                    }.copy(nullable = type.isMarkedNullable)
                }
        }

    private fun resolve(arg: KSTypeArgument): TypeName =
        when (arg.variance) {
            Variance.STAR -> WildcardTypeName.producerOf(ANY.copy(nullable = true))
            Variance.COVARIANT ->
                WildcardTypeName.producerOf(resolve(arg.type!!.fastResolve()))
            Variance.CONTRAVARIANT ->
                WildcardTypeName.consumerOf(resolve(arg.type!!.fastResolve()))
            else ->
                resolve(arg.type!!.fastResolve())
        }

    data class Result(
        val arguments: List<KSTypeArgument>,
        val argumentTypeNames: List<TypeName>
    )

    private class Finished(
        val result: Result
    ) : Exception()
}